File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/AdapterDataChangeObservable.java
package com.jakewharton.rxbinding2.widget;

import android.database.DataSetObserver;
import android.widget.Adapter;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class AdapterDataChangeObservable<T extends Adapter> extends InitialValueObservable<T> {
  private final T adapter;

  AdapterDataChangeObservable(T adapter) {
    this.adapter = adapter;
  }

  @Override protected void subscribeListener(Observer<? super T> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    ObserverDisposable<T> disposableDataSetObserver = new ObserverDisposable<>(adapter, observer);
    adapter.registerDataSetObserver(disposableDataSetObserver.dataSetObserver);
    observer.onSubscribe(disposableDataSetObserver);
  }

  @Override protected T getInitialValue() {
    return adapter;
  }

  static final class ObserverDisposable<T extends Adapter> extends MainThreadDisposable {
    private final T adapter;
    private final DataSetObserver dataSetObserver;

    ObserverDisposable(final T adapter, final Observer<? super T> observer) {
      this.adapter = adapter;
      this.dataSetObserver = new DataSetObserver() {
        @Override public void onChanged() {
          if (!isDisposed()) {
            observer.onNext(adapter);
          }
        }
      };
    }

    @Override protected void onDispose() {
      adapter.unregisterDataSetObserver(dataSetObserver);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/AdapterDataChangeObservable.kt
@file:JvmName("RxAdapter")
@file:JvmMultifileClass

package com.jakewharton.rxbinding2.widget

import android.database.DataSetObserver
import android.widget.Adapter
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding2.InitialValueObservable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding2.internal.checkMainThread

/**
 * Create an observable of data change events for `adapter`.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun <T : Adapter> T.dataChanges(): InitialValueObservable<T> {
  return AdapterDataChangeObservable(this)
}

private class AdapterDataChangeObservable<T : Adapter>(
  private val adapter: T
) : InitialValueObservable<T>() {

  override fun subscribeListener(observer: Observer<in T>) {
    if (!checkMainThread(observer)) {
      return
    }
    val disposableDataSetObserver = ObserverDisposable(initialValue, observer)
    initialValue.registerDataSetObserver(disposableDataSetObserver.dataSetObserver)
    observer.onSubscribe(disposableDataSetObserver)
  }

  override val initialValue get() = adapter

  private class ObserverDisposable<T : Adapter>(
    private val adapter: T,
    observer: Observer<in T>
  ) : MainThreadDisposable() {
    @JvmField val dataSetObserver = object : DataSetObserver() {
      override fun onChanged() {
        if (!isDisposed) {
          observer.onNext(adapter)
        }
      }
    }

    override fun onDispose() {
      adapter.unregisterDataSetObserver(dataSetObserver)
    }
  }
}
